exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/RemoteFunctionRegistry.java

```diff
@@ -0,0 +1,185 @@
+  private TransientStore<String> jars;
+
+  public RemoteFunctionRegistry(TransientStoreListener unregistrationListener) {
+    this.unregistrationListener = unregistrationListener;
+  }
+
+  public void init(DrillConfig config, PersistentStoreProvider storeProvider, ClusterCoordinator coordinator) {
+    try {
+      PersistentStoreConfig<Registry> registrationConfig = PersistentStoreConfig
+          .newProtoBuilder(SchemaUserBitShared.Registry.WRITE, SchemaUserBitShared.Registry.MERGE)
+          .name("udf")
+          .persist()
+          .build();
+      registry = storeProvider.getOrCreateStore(registrationConfig);
+      registry.putIfAbsent(REGISTRY, Registry.getDefaultInstance());
+    } catch (StoreException e) {
+      throw new DrillRuntimeException("Failure while loading remote registry.", e);
+    }
+
+    TransientStoreConfig<String> unregistrationConfig = TransientStoreConfig.
+        newJacksonBuilder(mapper, String.class).name("udf/unregister").build();
+    unregistration = coordinator.getOrCreateTransientStore(unregistrationConfig);
+    unregistration.addListener(unregistrationListener);
+
+    TransientStoreConfig<String> jarsConfig = TransientStoreConfig.
+        newJacksonBuilder(mapper, String.class).name("udf/jars").build();
+    jars = coordinator.getOrCreateTransientStore(jarsConfig);
+
+    this.retryTimes = config.getInt(ExecConstants.UDF_RETRY_TIMES);
+
+    Configuration conf = new Configuration();
+    if (config.hasPath(ExecConstants.UDF_DIRECTORY_FS)) {
+      conf.set(FileSystem.FS_DEFAULT_NAME_KEY, config.getString(ExecConstants.UDF_DIRECTORY_FS));
+    }
+
+    try {
+      this.fs = FileSystem.get(conf);
+      String base = config.getString(ExecConstants.UDF_DIRECTORY_BASE);
+      this.registryArea = createArea(fs, base, config.getString(ExecConstants.UDF_DIRECTORY_REGISTRY));
+      this.stagingArea = createArea(fs, base, config.getString(ExecConstants.UDF_DIRECTORY_STAGING));
+      this.tmpArea = createArea(fs, base, config.getString(ExecConstants.UDF_DIRECTORY_TMP));
+    } catch (IOException e) {
+      throw new DrillRuntimeException("Error during areas creation for remote function registry.", e);
+    }
+  }
+
+  public Registry getRegistry() {
+    return registry.get(REGISTRY);
+  }
+
+  public Registry getRegistry(DataChangeVersion version) {
+    return registry.get(REGISTRY, version);
+  }
+
+  public void updateRegistry(Registry registryContent, DataChangeVersion version) throws VersionMismatchException {
+    registry.put(REGISTRY, registryContent, version);
+  }
+
+  public void submitForUnregistration(String jar) {
+    putIfAbsent(unregistration, jar, jar);
+  }
+
+  public void finishUnregistration(String jar) {
+    remove(unregistration, jar);
+  }
+
+  public String addToJars(String jar, Action action) {
+    return putIfAbsent(jars, jar, action.toString());
+  }
+
+  public void removeFromJars(String jar) {
+    remove(jars, jar);
+  }
+
+  public int getRetryTimes() {
```

Comment:
Is this needed in the public interface? Or, is this just a private implementation detail? Perhaps needed for unit testing?